src/AdoNetCore.AseClient/Internal/InternalConnectionFactory.cs

```diff
@@ -190,20 +191,45 @@ private bool UserCertificateValidationCallback(object sender, X509Certificate se
                 return false;
             }
 
-            var untrustedRootChainStatusFlags = false;
-            var otherChainStatusFlags = false;
-
-            // We're not concerned with UntrustedRoot errors as we verify that below.
+            var mergedStatusFlags = X509ChainStatusFlags.NoError;
             foreach (var status in chain.ChainStatus)
             {
-                untrustedRootChainStatusFlags |= (status.Status & X509ChainStatusFlags.UntrustedRoot) == X509ChainStatusFlags.UntrustedRoot;
-                otherChainStatusFlags |= (status.Status & ~X509ChainStatusFlags.UntrustedRoot) != X509ChainStatusFlags.NoError;
+                mergedStatusFlags |= status.Status;
+            }
+
+            var trustedCerts = LoadTrustedFile(_parameters.TrustedFile);
+            if (trustedCerts == null)
+            {
+                Logger.Instance?.WriteLine($"{nameof(InternalConnectionFactory)}.{nameof(UserCertificateValidationCallback)} secure connection failed due to missing TrustedFile parameter.");
+                return false;
+            }
 
-                if (otherChainStatusFlags)
+#if !(NETCOREAPP1_0 || NETCOREAPP1_1) // these frameworks do not have the following X509Certificate2 constructor...
+            // sometimes the chain policy is only a partial chain because it doesn't include a self signed root?
+            if ((mergedStatusFlags & X509ChainStatusFlags.PartialChain) == X509ChainStatusFlags.PartialChain)
+            {
+                // attempt to resolve a partial root by rebuilding the cert chain including the certs from the trusted file
+                chain.ChainPolicy.ExtraStore.AddRange(trustedCerts);
+                if (chain.Build(new X509Certificate2(serverCertificate)))
                 {
-                    Logger.Instance?.WriteLine($"{nameof(InternalConnectionFactory)}.{nameof(UserCertificateValidationCallback)} secure connection failed due to chain status: {status.Status}");
-                    return false;
+                    // Chain validated with extra roots added; accept it
+                    return true;
                 }
+                mergedStatusFlags = X509ChainStatusFlags.NoError;
+                foreach (var status in chain.ChainStatus)
+                {
+                    mergedStatusFlags |= status.Status;
+                }
+            }
+#endif
+
+            var untrustedRootChainStatusFlags = (mergedStatusFlags & X509ChainStatusFlags.UntrustedRoot) == X509ChainStatusFlags.UntrustedRoot;
+            var otherChainStatusFlags = (mergedStatusFlags & ~X509ChainStatusFlags.UntrustedRoot) != X509ChainStatusFlags.NoError;
+
+            if (otherChainStatusFlags)
```

Comment:
If the chain had an error somewhere in the middle, eg [ServerCert, IntermediateCert, **ExpiredCert**, RootCert], we should probably still check if ServerCert or IntermediateCert are directly in the trusted file; the user has explicitly trusted a cert that is used earlier in the chain than where the error occurred.

I think this is also a problem in the original code too, so this pr is still an improvement either way.

Reply:
We'll leave this out of scope for the PR. It's not clear-cut whether this would be a good or bad thing. @SSMKittel if the need is strong, create an issue and we can resolve it there.